### PR TITLE
1.7/1303 revisit streaming/decrypting the MVL (V/EE) 

### PR DESCRIPTION
**Pull request by Hoolean**
# Progress

This PR implements the improvements suggested in #297, serving the Master Voucher Log with a streaming approach at every stage. As a result, memory usage rests comfortably at only a few megabytes and is now constant for any log size.

## New Flow

All spreadsheets are now compressed prior to encryption, as opposed to vice-versa.

Encryption and decryption occurs on output and input streams respectively, avoiding the need for either of these to hold the entire file contents in memory at any given time. This is achieved with:
* A stream wrapper for the log generation process
* Direct usage of the decryption streaming primitive in the log serving process.

### Log Creation

```
Database -> Spreadsheet -> ZIP Stream -> Secret Stream Wrapper -> File on Disk
```

### Log Serving

```
File on Disk -> Decryption Stream -> HTTP Response Stream
```

## Caveats
* The log's file size is fiddly to calculate prior to decryption, resulting in a download progress bar that lacks an estimation of progress and time remaining.
* A custom error page cannot be shown after the system has started serving the log, as an ongoing HTTP streaming response cannot be modified very semantically.
* The encryption library is not integrated into Laravel (but is part of the PHP standard library, as of PHP 7.2).

## Features
* The approach is scalable and should not require future optimisation.
* The log file is now compressed to a far smaller size, prior to encryption, reducing its use of memory, bandwidth and storage resources.
* The implementation of streaming encryption is highly reusable and could help reduce memory usage in other processes (including the generation of the file).

### CODE REVIEW

File: app/Wrappers/SecretStreamWrapper.php [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    public static function getKey() {
+        $key = config('app.key');
+
+        // Sometimes, the key is encoded in base64. Decrypt it if this is the case.
+        if (Str::startsWith($key, 'base64:')) {
+            $key = base64_decode(substr($key, 7));
+        }
+
+        // Check that the key is the write length.
+        // TODO : Can we guarantee that app.key will be configured as 32 random bytes?
```
Comment (charlesstrange2) [inline app/Wrappers/SecretStreamWrapper.php]: If people use `artisan key:generate` then yes, it should be.

File: app/Wrappers/SecretStreamWrapper.php [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+        // Flush out the final message, tagged as such.
+        if (strlen($this->buffer) > 0) {
+            $encrypted = sodium_crypto_secretstream_xchacha20poly1305_push(
+                $this->stream,
+                $this->buffer,
+                '', // This is an inactive placeholder.
+                SODIUM_CRYPTO_SECRETSTREAM_XCHACHA20POLY1305_TAG_FINAL
+            );
+            fwrite($this->wrapped, $encrypted); // TODO : Can an error here be caught somehow?
```
Comment (charlesstrange2) [inline app/Wrappers/SecretStreamWrapper.php]: frwite, being a fairly basic function returns falsey (false or, possibly 0) as an error state. not terribly useful

Comment (Hoolean) [reply app/Wrappers/SecretStreamWrapper.php]: From testing, after writing the comment, it seemed that Laravel was eager to treat these errors pretty seriously without having to spot them manually. The error would be spotted by the single user on serving the file as well, so I think I'll leave this uncaught :+1:

File: app/Wrappers/SecretStreamWrapper.php (hunk 1 above)
Comment (Hoolean) [reply app/Wrappers/SecretStreamWrapper.php]: I don't see much of a risk of anything insecure but also the right length being used here. We'll get a nice error spat out if the length is wrong too, so I think we should be able to leave this as is :+1:

**Pull request merged**